### DISCUSSION TITLE
Authentication JWT bad username or password after disconnecting

### DISCUSSION THREAD

**Discussion by MacherelR** [emqx/emqx#12043] [Q&A]
Hi, 
I deployed an emqx cluster with the jwt authentication plugin. And I have a custom authentication server that shares a secret with the emqx broker and uses it to create JWT tokens on demand.
Now my other app is asking for a token, then uses it to connect to emqx broker and this works well on first install (app is installed in the cluster using helm install). When declaring a client using paho mqtt python library, I use the clean_session=False to make my session persistent.
Now sometimes (not always, that's why it intrigues me), if I uninstall my app (helm uninstall) and then reinstalls it it tries to connect again to the broker using the same client ID and same username but a different token (freshly asked from the auth service) and the connection gets refused with rc = 4 (bad username or password). Now what intrigues me is that this doesn't always happen, sometimes the app is able to connect flawlessly and sometimes not. It looks like this happens after some uptime of broker but still can't figure out why ? 
Has anyone faced something similar ?

**Comment by SergeTupchiy** [comment]
Hi @MacherelR, 
Are you literally using some JWT plugin or EMQX builtin JWT authentication (https://www.emqx.io/docs/en/latest/access-control/authn/jwt.html#jwt-authentication)?
The latter option is recommended and must be available for you, as judging by your previous question you are using EMQX 5.1.6.

**Reply by MacherelR** [reply]
Hello @SergeTupchiy,
I'm using EMQX builtin configured as follows in the deployment:
~~~
authentication = [{
            from = password
            mechanism = jwt
            enable = true
            secret = <SECRET>
            secret_base64_encoded = false
            algorithm = hmac-based
            use_jwks = false
            verify_claims {
               username = "${username}"
            }
            acl_claim_name = "acl"
         }]
~~~
It just seems that sometimes after a long time on emqx refuses accesses and I can't figure out why

**Comment by SergeTupchiy** [comment]
@MacherelR,
Do you use any of "exp" (expiration time), "nbf" (not before time) , "iat" (issued at time) JWT claims? 
EMQX validates them if they are provided. 
If a token is rejected after a long time it looks like it can be simply expired. 
I'd recommend enabling tracing: https://www.emqx.io/docs/en/latest/dashboard/diagnose.html#log-trace and trying to catch this case.
The tracing facility should log some extra details.

**Reply by MacherelR** [reply]
I use expiration time and it's valid on the tokens that  are rejected. I also use iat but not nbf.
What could be the problem with iat ? 

I'll try enabling log trace and reproduce the error !

**Comment by SergeTupchiy** [comment]
@MacherelR,
EMQX checks that iat is not greater than the current system time (seconds), i. e. it validates that "iat" is not in the future.

**Comment by vetinary** [comment]
Did you find the answer for your question? I'm experiencing the same problem and can't figure out how to fix that issue

**Reply by zmstone** [reply]
> Did you find the answer for your question? I'm experiencing the same problem and can't figure out how to fix that issue

Hi @vetinary 
Which version are you using?
Please check if the token's "exp" (expiration time), "nbf" (not before time) , "iat" (issued at time)  are valid according to the EMQX node's system time.
If the time stamps are valid, please create an issue with:
1. EMQX authentication configuration
2. An example token
3. Trace (created using client ID) logs

**Reply by vetinary** [reply]
Hi @zmstone 
I'll prepare requested information. Before I create a separate issue with detailed information, I'll describe the problem in general.
I create token using iat and exp properties. Let's say, iat is "now" and exp is "now + 60*60". And everything works great. Each time I open the page, the new token is created to make it valid for a long term since every page reload. 
The problem appears when time of exp of the first token generated for username is reached: no matter which exp property is in new generated token, I get "reason: chain_result, result: {stop,{error,bad_username_or_password}}, tag: AUTHN" in emqx console. If I restart emqx, everything works great, however, again, when exp of the first generated token is reached, I can't connect EMQX with a new generated token. The only way to overcome it is to create token with very big exp property.

EMQX version 5.1.2 on MacOS.

I'll make more tests and create a separate issue later.

**Reply by savonarola** [reply]
Hello!

I tried to reproduce with the following setup.

We have `./emqx.conf`:

```
node {
  name = "emqx@127.0.0.1"
  cookie = "emqxsecretcookie"
  data_dir = "data"
}

cluster {
  name = emqxcl
  discovery_strategy = manual
}

dashboard {
    listeners.http {
        bind = 18083
    }
}

authentication = [{
            from = password
            mechanism = jwt
            enable = true
            secret = secret
            secret_base64_encoded = false
            algorithm = hmac-based
            use_jwks = false
            verify_claims {
               username = "${username}"
            }
            acl_claim_name = "acl"
         }]
```

We run emqx with the command:

```
docker run -d --name emqx \
  -p 18083:18083 -p 1883:1883 \
  -v $(pwd)/emqx.conf:/opt/emqx/etc/emqx.conf \
  emqx:5.1.2
```

We check that we fail to connect without a proper password:
```
mosquitto_sub -d -t 't1' -u u2 -P wrongpass -i u3
Client u3 sending CONNECT
Client u3 received CONNACK (5)
Connection error: Connection Refused: not authorised.
Client u3 sending DISCONNECT
```

Now we generate jwt with 20s exp and connect

```
>pass=$(python3 -c 'import time; import jwt; t = int(time.time()); print(jwt.encode({"username": "u2", "iat": t, "exp": t + 20}, "secret", algorithm="HS256"))'); echo $pass
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InUyIiwiaWF0IjoxNzIwNTQwNjkwLCJleHAiOjE3MjA1NDA3MTB9.iIB6XI2i1JmuvQIipU46f8qzjYLkm7xvl3liefGX-rU

>mosquitto_sub -d -t 't1' -u u2 -P $pass -i u2 -V mqttv5 -x 3600 -c
Client u2 sending CONNECT
Client u2 received CONNACK (0)
Client u2 sending SUBSCRIBE (Mid: 1, Topic: t1, QoS: 0, Options: 0x00)
Client u2 received SUBACK
Subscribed (mid: 1): 0
```

We do not stop the connection and wait  >20s.
**In a separate shell**, we try to connect with the same JWT and fail

```
mosquitto_sub -d -t 't1' -u u2 -P eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InUyIiwiaWF0IjoxNzIwNTQwNjkwLCJleHAiOjE3MjA1NDA3MTB9.iIB6XI2i1JmuvQIipU46f8qzjYLkm7xvl3liefGX-rU -i u2 -V mqttv5 -x 3600 -c
Client u2 sending CONNECT
Client u2 received CONNACK (134)
Connection error: Bad User Name or Password
Client u2 sending DISCONNECT
```

In the **same second shell** we regenerate JWT and connect:
```
>pass=$(python3 -c 'import time; import jwt; t = int(time.time()); print(jwt.encode({"username": "u2", "iat": t, "exp": t + 20}, "secret", algorithm="HS256"))'); echo $pass
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InUyIiwiaWF0IjoxNzIwNTQwODM2LCJleHAiOjE3MjA1NDA4NTZ9.UHFS_sNBrdBCBppBP-hPcF3heQjw3tbj5LV3lBCi8Lw
>mosquitto_sub -d -t 't1' -u u2 -P $pass -i u2 -V mqttv5 -x 3600 -c
Client u2 sending CONNECT
Client u2 received CONNACK (0)
Client u2 sending SUBSCRIBE (Mid: 1, Topic: t1, QoS: 0, Options: 0x00)
Client u2 received SUBACK
Subscribed (mid: 1): 0
```

In the first shell
```
...
Received DISCONNECT (142)
Client u2 sending CONNECT
Client u2 received CONNACK (134)
Connection error: Bad User Name or Password
Client u2 sending DISCONNECT
```
We see that the client was kicked. 

The bug cannot be reproduced.

So,
* Could you verify the scenario? Does it match the use case where the bug appears?
* Could we modify it to reproduce the bug?
* Please note, that in the scenario the kicked client tried to reconnect immediately but, as the JWT was old, it couldn't. This may be a common behaviour resulting to connection errors in logs.
* Also, for the failing client with valid JWT, could you [enable traces](https://docs.emqx.com/en/emqx/latest/observability/tracer.html) and collect the logs?
* Do you use acl rules in JWT?

**Reply by vetinary** [reply]
Hi @savonarola 

Thank you for your time spent on my issue. 
Looks like the problem is in the way Mac OS saves state of running processes when it goes to sleep mode. 
When I close my laptop, Mac OS and open it several minutes later, EMQX stops accepting tokens, no mater which `exp` property they have in claim. And I have to restart EMQX server to make it accept tokens. But if I don't put Mac OS to sleep mode, everything works fine. 
I have just reproduced it several times, closing my laptop and then opening it again after 5 minutes and trying to create new token and connect to EMQX — connection fails with error `result: {stop,{error,bad_username_or_password}}`

**Reply by vetinary** [reply]
Looks like it was an issue with old EMQX installed. I updated EMQX to version 5.7.1 and now everything works like a charm! 
Thank you for your help
